util/base_parser.h

```diff
@@ -21,9 +21,11 @@ class BaseParser {
 public:
     constexpr explicit BaseParser(std::string_view input) : input_{input} {}
 
-    constexpr char peek() const { return input_[pos_]; }
+    constexpr char peek() const { return (is_eof() ? 0xff : input_[pos_]); }
```

Comment:
It was a bitch, but I re-implemented it with an optional and it seems to work correctly and it looks safe everywhere it's used. I rewrote `peek()` instead of adding a new function, because there's no way for me to know when to call the "try" version.